primer/src/Primer/App.hs

```diff
@@ -692,14 +692,39 @@ applyProgAction prog = \case
       ( m{moduleTypes = tydefs'}
       , Just $ SelectionTypeDef $ TypeDefSelection tc Nothing
       )
-  DeleteTypeDef d -> editModuleCross (qualifiedModule d) prog $ \(m, ms) ->
-    case moduleTypesQualified m Map.!? d of
-      Nothing -> throwError $ TypeDefNotFound d
-      Just (TypeDefPrim _) -> throwError $ TypeDefIsPrim d
-      Just (TypeDefAST td) -> do
-        checkTypeNotInUse d td $ m : ms
-        let m' = m{moduleTypes = Map.delete (baseName d) (moduleTypes m)}
-        pure (m' : ms, Nothing)
+  DeleteTypeDef d -> editModuleOfCrossType (Just d) prog $ \(m, ms) defName def -> do
+    let updateExpr = \case
+          Con _ c _ | c `elem` map valConName (astTypeDefConstructors def) -> emptyHole
+          e -> pure e
+    ms' <-
+      ((m & over #moduleTypes (Map.delete defName)) : ms)
+        & ( traverseOf
+              (traversed % #moduleDefs % traversed % #_DefAST)
+              ( traverseOf
+                  #astDefExpr
+                  ( transformM (traverseOf typesInExpr updateType)
+                      <=< transformM updateExpr
+                  )
+                  <=< traverseOf #astDefType updateType
+              )
+              <=< traverseOf
+                ( traversed
+                    % #moduleTypes
+                    % traversed
+                    % #_TypeDefAST
+                    % #astTypeDefConstructors
+                    % traversed
+                    % #valConArgs
+                    % traversed
+                )
+                updateType
+          )
+    ms'' <- runFullTCPass (progSmartHoles prog) (progImports prog) ms'
+    pure (ms'', Nothing)
+    where
+      updateType = transformM \case
+        TCon _ n | n == d -> tEmptyHole
+        e -> pure e
```

Comment:
It is somewhat odd to have `updateExpr` defined above in a `let`, but `updateType` defined below in a `where`. Any particular reason?

Reply:
Agreed, but we mostly use `where` in the other cases, but `updateExpr` can't go there because then `def` would be out of scope. It's a bit weird, but I don't think it's worth changing.

Reply:
Agreed, but we mostly use `where` in the other cases, but `updateExpr` can't go there because then `def` would be out of scope. It's a bit weird, but I don't think it's worth changing.